### PR TITLE
ENH: add in extension dtype registry

### PR DESCRIPTION
**Pull request by jreback**
precursor to #21160

### CODE REVIEW

File: doc/source/whatsnew/v0.24.0.txt [hunk 1]
```diff
@@ -14,14 +14,22 @@ Other Enhancements
...
 Backwards incompatible API changes
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
+.. _whatsnew_0240.api.extension:
+
+ExtensionType Changes
+^^^^^^^^^^^^^^^^^^^^^
+
+- ``ExtensionArray`` has gained the abstract methods ``.dropna()`` and ``.append()``, and attribute ``array_type`` (:issue:`21185`)
+- ``ExtensionDtype`` has gained the ability to instantiate from string dtypes, e.g. ``decimal`` would instaniate a registered ``DecimalDtype`` (:issue:`21185`)
```
Comment (jbrockmendel) [inline doc/source/whatsnew/v0.24.0.txt]: typo instantiate

**Review by Dr-Irv — COMMENTED** [review]
It seems to me that there are changes in this PR that go beyond just adding in the extension dtype registry.  Wasn't your goal to just make this PR about the registry?

**Comment by jreback** [comment]
split out ops to a separate PR #21191

File: pandas/core/dtypes/dtypes.py [hunk 2]
```diff
@@ -2,12 +2,70 @@
...
+    These are tried in order for inference.
+    """
+    dtypes = OrderedDict()
+
+    @classmethod
+    def register(self, dtype, constructor=None):
+        """
+        Parameters
+        ----------
+        dtype : PandasExtension Dtype
```
Comment (jorisvandenbossche) [inline pandas/core/dtypes/dtypes.py]: Can you document constructor parameter? 
and "PandasExtension Dtype" -> "ExtensionDtype"

File: pandas/core/dtypes/dtypes.py [hunk 3]
```diff
@@ -2,12 +2,70 @@
...
+        """
+        if not issubclass(dtype, (PandasExtensionDtype, ExtensionDtype)):
```
Comment (jorisvandenbossche) [inline pandas/core/dtypes/dtypes.py]: Shouldn't ExtensionDtype be enough? (our internals one subclass both I think?)

File: pandas/core/dtypes/dtypes.py [hunk 4]
```diff
@@ -564,6 +622,17 @@ def construct_from_string(cls, string):
...
+    @classmethod
+    def construct_from_string_strict(cls, string):
+        """
+        Strict construction from a string, raise a TypeError if not
+        possible
+        """
+        if string.startswith('period[') or string.startswith('Period['):
+            # do not parse string like U as period[U]
+            return PeriodDtype.construct_from_string(string)
+        raise TypeError("could not construct PeriodDtype")
```
Comment (jorisvandenbossche) [inline pandas/core/dtypes/dtypes.py]: Should we consider rather changing `construct_from_string` itself? As those don't really adhere to the ExtensionDtype requirements I think?

If we keep both, I would make this one private.

File: pandas/tests/extension/base/ops.py [hunk 5]
```diff
@@ -0,0 +1,6 @@
+from .base import BaseExtensionTests
+
+
+class BaseOpsTests(BaseExtensionTests):
+    """Various Series and DataFrame ops methos."""
+    pass
```
Comment (jorisvandenbossche) [inline pandas/tests/extension/base/ops.py]: This is a left-over from splitting it from the previous PR?

**Review by jorisvandenbossche — COMMENTED** [review]
Didn't look at all the tests yet.

In addition:

- We need to discuss what is the exact contract for this registry and `array_type`: where is it used, and how is it used? 
  I think now it is used in constructors or astype if you specify a (string) dtype? But then it calls the class constructor of the extension array? Shouldn't this rather be the `_from_sequence` method ?
- Can you add some explanation of the registry to the extension array documentation?
- I think this also gives an enhancement for the existing types (interval, datetimetz, period, ..) that you can now specify string dtypes? (since they are added to the registry) In that case, can you add tests for this and a whatsnew notice?

File: doc/source/whatsnew/v0.24.0.txt [hunk 6]
```diff
@@ -14,14 +14,22 @@ Other Enhancements
...
 
 Backwards incompatible API changes
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
+.. _whatsnew_0240.api.extension:
+
+ExtensionType Changes
+^^^^^^^^^^^^^^^^^^^^^
+
+- ``ExtensionArray`` has gained the abstract methods ``.dropna()`` and ``.append()``, and attribute ``array_type`` (:issue:`21185`)
```
[the remark this answers is not in the document]
Comment (jreback) [reply doc/source/whatsnew/v0.24.0.txt]: no, these are necessary for followons. The api is not fleshed out enough.

File: pandas/core/arrays/base.py [hunk 7]
```diff
@@ -82,14 +85,16 @@ class ExtensionArray(object):
...
+    def _from_sequence(cls, scalars, copy=False):
         """Construct a new ExtensionArray from a sequence of scalars.
 
         Parameters
         ----------
         scalars : Sequence
             Each element will be an instance of the scalar type for this
             array, ``cls.dtype.type``.
+        copy : boolean, default True
+            if True, copy the underlying data
```
[the remark this answers is not in the document]
Comment (jreback) [reply pandas/core/arrays/base.py]: same guarantee with copy we have already, if its True, then copy if possible.

File: pandas/core/dtypes/cast.py [hunk 8]
```diff
@@ -647,6 +647,11 @@ def conv(r, dtype):
 def astype_nansafe(arr, dtype, copy=True):
     """ return a view if copy is False, but
         need to be very careful as the result shape could change! """
+
+    # dispatch on extension dtype if needed
+    if is_extension_array_dtype(dtype):
+        return dtype.array_type._from_sequence(arr, copy=copy)
```
[the remark this answers is not in the document]
Comment (jreback) [reply pandas/core/dtypes/cast.py]: this validation is a contract of the Array itself

File: pandas/core/dtypes/dtypes.py (hunk 3 above)
Comment (jreback) [reply pandas/core/dtypes/dtypes.py]: not yet, are extension dtypes are not all ExtensiDtypes yet (this is why we have the whole PandasExtensionDtype bizness)

File: pandas/core/dtypes/dtypes.py (hunk 4 above)
Comment (jreback) [reply pandas/core/dtypes/dtypes.py]: done

**Comment by jreback** [comment]
all fixed up @jorisvandenbossche  (except for docs on Registry in Extension section)

**Comment by jreback** [comment]
see if can make this dispatch to create_array_type as a function

**Comment by jreback** [comment]
updated docs

@jorisvandenbossche @TomAugspurger if any final comments.

File: pandas/tests/extension/base/constructors.py [hunk 9]
```diff
@@ -45,3 +46,14 @@ def test_series_given_mismatched_index_raises(self, data):
         msg = 'Length of passed values is 3, index implies 5'
         with tm.assert_raises_regex(ValueError, msg):
             pd.Series(data[:3], index=[0, 1, 2, 3, 4])
+
+    def test_from_dtype(self, data):
+        # construct from our dtype & string dtype
+        dtype = data.dtype
+
+        expected = pd.Series(data)
+        result = pd.Series(np.array(data), dtype=dtype)
```
Comment (jorisvandenbossche) [inline pandas/tests/extension/base/constructors.py]: Do we have a restriction on what the result of `__array__` should be? Maybe not necessarily the scalars? 

So maybe better to do `list(data)` or `data.astype(object)`

File: pandas/tests/extension/category/test_categorical.py [hunk 10]
```diff
@@ -52,7 +52,9 @@ def data_for_grouping():
 
 
 class TestDtype(base.BaseDtypeTests):
-    pass
+
+    def test_array_type_with_arg(self, data, dtype):
+        assert dtype.construct_array_type('foo') is Categorical
```
Comment (jorisvandenbossche) [inline pandas/tests/extension/category/test_categorical.py]: Why should this not raise?

File: pandas/core/dtypes/base.py [hunk 11]
```diff
@@ -156,6 +161,22 @@ def name(self):
         """
         raise AbstractMethodError(self)
 
+    @classmethod
+    def construct_array_type(cls, array=None):
+        """Return the array type associated with this dtype
+
+        Parameters
+        ----------
+        array : array-like, optional
```
Comment (jorisvandenbossche) [inline pandas/core/dtypes/base.py]: Do we actually have a use case for this at the moment ?

File: pandas/tests/extension/decimal/array.py [hunk 12]
```diff
@@ -27,7 +41,7 @@ def construct_from_string(cls, string):
 class DecimalArray(ExtensionArray):
     dtype = DecimalDtype()
 
-    def __init__(self, values):
+    def __init__(self, values, copy=False):
```
Comment (jorisvandenbossche) [inline pandas/tests/extension/decimal/array.py]: this change should not be needed?

File: pandas/tests/extension/decimal/test_decimal.py [hunk 13]
```diff
@@ -92,15 +92,21 @@ def assert_frame_equal(self, left, right, *args, **kwargs):
...
     pass
 
 
 class TestConstructors(BaseDecimal, base.BaseConstructorsTests):
-    pass
+
+    @pytest.mark.xfail(reason="not implemented constructor from dtype")
+    def test_from_dtype(self, data):
+        # construct from our dtype & string dtype
+        pass
```
Comment (jorisvandenbossche) [inline pandas/tests/extension/decimal/test_decimal.py]: Can you actually implement this? (only change needed would be to register DecimalDtype I think?) 
As that way we actually have a test for external dtypes registering?

File: pandas/tests/extension/decimal/test_decimal.py [hunk 14]
```diff
@@ -147,6 +153,10 @@ class TestGroupby(BaseDecimal, base.BaseGroupbyTests):
     pass
 
 
+# TODO(extension)
+@pytest.mark.xfail(reason=(
+    "raising AssertionError as this is not implemented, "
+    "though easy enough to do"))
```
Comment (jorisvandenbossche) [inline pandas/tests/extension/decimal/test_decimal.py]: I think you can change the AssertionError to a ValueError in the code, and then we can still test this.

File: pandas/tests/extension/json/test_json.py [hunk 15]
```diff
@@ -130,7 +132,11 @@ def test_custom_asserts(self):
 
 
 class TestConstructors(BaseJSON, base.BaseConstructorsTests):
-    pass
+
+    @pytest.mark.xfail(reason="not implemented constructor from dtype")
+    def test_from_dtype(self, data):
+        # construct from our dtype & string dtype
+        pass
```
Comment (jorisvandenbossche) [inline pandas/tests/extension/json/test_json.py]: Can we test here that an appropriate error is raised (instead of skipping the test) ?

File: pandas/tests/dtypes/test_dtypes.py [hunk 16]
```diff
@@ -767,3 +773,24 @@ def test_update_dtype_errors(self, bad_dtype):
...
+     PeriodDtype, IntervalDtype])
+def test_registry(dtype):
+    assert dtype in registry.dtypes
+
+
+@pytest.mark.parametrize(
+    'dtype, expected',
+    [('int64', None),
+     ('interval', IntervalDtype()),
+     ('interval[int64]', IntervalDtype()),
```
[the remark this answers is not in the document]
Comment (jorisvandenbossche) [reply pandas/tests/dtypes/test_dtypes.py]: (although that is maybe covered with the period or datetime64 one below. Is `'interval[datetime64[ns]]'` already tested for IntervalDtype?)

File: pandas/core/dtypes/base.py [hunk 17]
```diff
@@ -109,6 +109,11 @@ class ExtensionDtype(_DtypeOpsMixin):
     * name
     * construct_from_string
 
+    Optionally one can override construct_array_type for construction
+    with the name of this dtype via the Registry
```
Comment (jorisvandenbossche) [inline pandas/core/dtypes/base.py]: I would give some additional explanation of what the Registry is, because now this is not explained here?

File: doc/source/whatsnew/v0.24.0.txt [hunk 18]
```diff
@@ -30,12 +30,22 @@ Datetimelike API Changes
 
 - For :class:`DatetimeIndex` and :class:`TimedeltaIndex` with non-``None`` ``freq`` attribute, addition or subtraction of integer-dtyped array or ``Index`` will return an object of the same class (:issue:`19959`)
 
+.. _whatsnew_0240.api.extension:
+
+ExtensionType Changes
+^^^^^^^^^^^^^^^^^^^^^
+
+- ``ExtensionArray`` has gained the abstract methods ``.dropna()`` and ``.append()`` (:issue:`21185`)
```
Comment (jorisvandenbossche) [inline doc/source/whatsnew/v0.24.0.txt]: As I said before, I am convinced we should necessarily add those methods (certainly append), so would personally leave this for another PR to discuss

**Review by jorisvandenbossche — COMMENTED** [review]
Added some more comments. 

From my previous review:

> We need to discuss what is the exact contract for this registry and array_type: where is it used, and how is it used?

Can you add some explanation about this to the docs/docstrings? (eg that it relies on `_from_sequence`)

File: pandas/core/dtypes/dtypes.py [hunk 19]
```diff
@@ -8,6 +8,64 @@
 from .base import ExtensionDtype, _DtypeOpsMixin
 
 
+class Registry(object):
```
Comment (TomAugspurger) [inline pandas/core/dtypes/dtypes.py]: Without looking at the uses yet, could we simplify this a by just allowing string lookup? Ideally, `registry` would be a simple class holding a dict mapping `dtype.name -> ExtensionDtype`.

File: pandas/core/indexes/interval.py [hunk 20]
```diff
@@ -800,7 +800,7 @@ def astype(self, dtype, copy=True):
     @cache_readonly
     def dtype(self):
         """Return the dtype object of the underlying data"""
-        return IntervalDtype.construct_from_string(str(self.left.dtype))
+        return IntervalDtype(str(self.left.dtype))
```
Comment (TomAugspurger) [inline pandas/core/indexes/interval.py]: `left.dtype.name`? I'm not sure when these differ, but using `.name` seems safer.

File: pandas/core/dtypes/dtypes.py (hunk 19 above)
Comment (jorisvandenbossche) [reply pandas/core/dtypes/dtypes.py]: I think the current code also supports finding the dtype for eg `'interval[int64]'` and not just `interval` (so parametrized versions of the strings)

Comment (TomAugspurger) [reply pandas/core/dtypes/dtypes.py]: Oh, and I suppose we want that to support `.astype('interval[int64]')`. That's fair...

**Comment by jreback** [comment]
ok all fixed up. @jorisvandenbossche @TomAugspurger if you want to look.

**Comment by jreback** [comment]
@Dr-Irv you don;t *need* to register the EA types, this is *only* for the purpose of translating a string dtype name -> EA dtype (e.g. 'categorical' -> Categorical), basically for convenience.

so sure you *could* have a collision, but then user would have to deal with that (this wouldn't show up on pandas because we won't have colliding names).

**Comment by jreback** [comment]
@Dr-Irv I had this an OrderdedDict but its actually not necessary, though maybe makes it more intuitve.

**Review by TomAugspurger — COMMENTED** [review]
Aside from adding EA.append, I'm +1 on this.

Given that NumPy doesn't implement it, I don't think we should either.

**Comment by jreback** [comment]
I had removed append FYI (it should not longer be in this PR)

**Comment by TomAugspurger** [comment]
I see it at https://github.com/pandas-dev/pandas/pull/21185/files#diff-99633b1041c2ea6283aa3d5bcf79e8cdR584

Maybe need to push?

**Comment by jreback** [comment]
pushed, yeah had it on the wrong branch.

**Comment by jreback** [comment]
any final comments. going to rebase.

**Comment by jreback** [comment]
@jorisvandenbossche @TomAugspurger

**Comment by jreback** [comment]
merging.
